### PR TITLE
slcore 642 replace hotspots methods with rpc

### PR DESCRIPTION
**Pull request by vojtech-suchy-sonarsource**

### CODE REVIEW

File: its/tests/src/test/java/its/SonarCloudTests.java [hunk 1]
```diff
@@ -526,27 +528,25 @@ void loadHotspotRuleDescription() throws Exception {
...
-      var projectBinding = new ProjectBinding(projectKey(PROJECT_KEY_JAVA_HOTSPOT), "", "ide");
-
-      engine.downloadAllServerHotspotsForFile(sonarcloudEndpointITOrg(), serverLauncher.getJavaImpl().getHttpClient(CONNECTION_ID), projectBinding,
-        "ide/src/main/java/foo/Foo.java", "master", null);
-
-      var serverHotspots = engine.getServerHotspots(projectBinding, "master", "ide/src/main/java/foo/Foo.java");
-      assertThat(serverHotspots)
-        .extracting("ruleKey", "message", "filePath", "textRange.startLine", "textRange.startLineOffset", "textRange.endLine", "textRange.endLineOffset", "status")
-        .containsExactly(tuple("java:S4792", "Make sure that this logger's configuration is safe.", "ide/src/main/java/foo/Foo.java", 9, 4, 9, 45, HotspotReviewStatus.TO_REVIEW));
+    void shouldMatchServerSecurityHotspots() throws ExecutionException, InterruptedException {
```
Comment (henryju) [inline its/tests/src/test/java/its/SonarCloudTests.java:531]: :+1:

File: its/tests/src/test/java/its/SonarQubeDeveloperEditionTests.java [hunk 2]
```diff
@@ -287,7 +321,11 @@ void shouldRaiseIssuesFromACustomRule() throws Exception {
       provisionProject(ORCHESTRATOR, projectKey, "Sample Java Custom");
       ORCHESTRATOR.getServer().restoreProfile(FileLocation.ofClasspath("/java-custom.xml"));
       ORCHESTRATOR.getServer().associateProjectToQualityProfile(projectKey, "java", "SonarLint IT Java Custom");
-      updateProject(engine, projectKey);
+
+      backend.getConfigurationService().didAddConfigurationScopes(new DidAddConfigurationScopesParams(
+        List.of(new ConfigurationScopeDto(CONFIG_SCOPE_ID, null, true, "projectName", new BindingConfigurationDto(CONNECTION_ID, projectKey, true)))));
+      await().atMost(20, SECONDS).untilAsserted(() -> assertThat(didSynchronizeConfigurationScopes).contains(CONFIG_SCOPE_ID));
+      await().untilAsserted(() -> assertThat(clientLogs.stream().anyMatch(s -> s.getMessage().equals("Stored project analyzer configuration"))).isTrue());
```
Comment (henryju) [inline its/tests/src/test/java/its/SonarQubeDeveloperEditionTests.java:328]: Maybe this duplicated block can be extracted in a separate method? Like `addConfigScopeAndWaitForAnalyzerSync`?

File: its/tests/src/test/java/its/SonarQubeDeveloperEditionTests.java [hunk 3]
```diff
@@ -763,15 +848,15 @@ void prepare() {
...
+      engine.stop();
     }
 
     @Test
     void shouldSyncBranchesFromServer() throws ExecutionException, InterruptedException {
       backend.getConfigurationService().didAddConfigurationScopes(new DidAddConfigurationScopesParams(
         List.of(new ConfigurationScopeDto(CONFIG_SCOPE_ID, null, true, "projectName", new BindingConfigurationDto(CONNECTION_ID, PROJECT_KEY, true)))));
 
-      await().untilAsserted(() -> assertThat(didSynchronizeConfigurationScopes).contains(CONFIG_SCOPE_ID));
+      await().atMost(20, SECONDS).untilAsserted(() -> assertThat(didSynchronizeConfigurationScopes).contains(CONFIG_SCOPE_ID));
```
Comment (henryju) [inline its/tests/src/test/java/its/SonarQubeDeveloperEditionTests.java:859]: Do we really need 20 seconds here???

File: its/tests/src/test/java/its/SonarQubeDeveloperEditionTests.java [hunk 4]
```diff
@@ -1047,16 +1128,19 @@ void start(TestInfo info) {
...
-        engine.stop(true);
+        engine.stop();
       } catch (Exception e) {
         // Ignore
       }
     }
 
     @Test
     @Tag(HOTSPOT_FEATURE_DISABLED)
+    @OnlyOnSonarQube(from = "9.7")
```
Comment (henryju) [inline its/tests/src/test/java/its/SonarQubeDeveloperEditionTests.java:1139]: What has changed for not supporting older SQ versions?

File: its/tests/src/test/java/its/SonarQubeDeveloperEditionTests.java [hunk 5]
```diff
@@ -1098,23 +1182,25 @@ private String getFirstHotspotKey(String projectKey) throws InvalidProtocolBuffe
     }
 
     @Test
+    @OnlyOnSonarQube(from = "9.7")
```
Comment (henryju) [inline its/tests/src/test/java/its/SonarQubeDeveloperEditionTests.java:1185]: What has changed for not supporting older SQ versions?

File: its/tests/src/test/java/its/SonarQubeDeveloperEditionTests.java [hunk 6]
```diff
@@ -1130,59 +1216,34 @@ void loadHotspotRuleDescription() throws Exception {
...
-      if (ORCHESTRATOR.getServer().version().isGreaterThanOrEquals(9, 7)
-        && !ORCHESTRATOR.getServer().version().isGreaterThanOrEquals(10, 1)) {
-        assertThat(serverHotspots)
-          .extracting("ruleKey", "message", "filePath", "textRange.startLine", "textRange.startLineOffset", "textRange.endLine", "textRange.endLineOffset", "status")
-          .containsExactly(
-            tuple("java:S4792", "Make sure that this logger's configuration is safe.", "ide/src/main/java/foo/Foo.java", 9, 4, 9, 45, HotspotReviewStatus.TO_REVIEW));
-      } else {
-        assertThat(serverHotspots).isEmpty();
-      }
+    @OnlyOnSonarQube(from = "9.7")
```
Comment (henryju) [inline its/tests/src/test/java/its/SonarQubeDeveloperEditionTests.java:1219]: What has changed for not supporting older SQ versions?

File: its/tests/src/test/java/its/SonarQubeDeveloperEditionTests.java [hunk 7]
```diff
@@ -1481,6 +1536,17 @@ public String matchSonarProjectBranch(String configurationScopeId, String mainBr
         return matchedBranchNameForProject == null ? super.matchSonarProjectBranch(configurationScopeId, mainBranchName, allBranchesNames, cancelChecker)
           : matchedBranchNameForProject;
       }
+
+      @Override
+      public void showSmartNotification(ShowSmartNotificationParams params) {
+        super.showSmartNotification(params);
```
Comment (henryju) [inline its/tests/src/test/java/its/SonarQubeDeveloperEditionTests.java:1542]: Is it useful?

File: its/tests/src/test/java/its/SonarQubeDeveloperEditionTests.java [hunk 8]
```diff
@@ -302,7 +340,11 @@ void shouldRaiseIssuesOnAPhpProject() throws Exception {
       provisionProject(ORCHESTRATOR, projectKey, "Sample PHP");
       ORCHESTRATOR.getServer().restoreProfile(FileLocation.ofClasspath("/php-sonarlint.xml"));
       ORCHESTRATOR.getServer().associateProjectToQualityProfile(projectKey, "php", "SonarLint IT PHP");
-      updateProject(engine, projectKey);
+
+      backend.getConfigurationService().didAddConfigurationScopes(new DidAddConfigurationScopesParams(
+        List.of(new ConfigurationScopeDto(CONFIG_SCOPE_ID, null, true, "projectName", new BindingConfigurationDto(CONNECTION_ID, projectKey, true)))));
+      await().atMost(20, SECONDS).untilAsserted(() -> assertThat(didSynchronizeConfigurationScopes).contains(CONFIG_SCOPE_ID));
+      await().untilAsserted(() -> assertThat(clientLogs.stream().anyMatch(s -> s.getMessage().equals("Stored project analyzer configuration"))).isTrue());
```
Comment (henryju) [inline its/tests/src/test/java/its/SonarQubeDeveloperEditionTests.java:347]: The fact that you need to await for logs is a red flag to me. We don't want real clients to do the same. Isn't the DidSynchronizeConfigurationScope enough?

Comment (vojtech-suchy-sonarsource) [reply its/tests/src/test/java/its/SonarQubeDeveloperEditionTests.java]: Checking for the `"Stored project analyzer configuration"` was added mostly for AnalysisTests. Without it, the engine.analyze() throws  (on cirrus, not reproducible locally) `java.nio.file.NoSuchFileException: ...../analyzer_config.pb` as it's still being updated (or at least the lock isn't released) even though we receive the DidSynchronizeConfigurationScope event.
I agree that waiting for a log is bad. I haven't had a better idea.

File: its/tests/src/test/java/its/SonarQubeDeveloperEditionTests.java (hunk 5 above)
Comment (vojtech-suchy-sonarsource) [reply its/tests/src/test/java/its/SonarQubeDeveloperEditionTests.java]: addressed in [4989ca3](https://github.com/SonarSource/sonarlint-core/pull/799/commits/4989ca37baf60cf950245a99149ed3079780e0fa)

File: its/tests/src/test/java/its/SonarQubeDeveloperEditionTests.java (hunk 4 above)
Comment (vojtech-suchy-sonarsource) [reply its/tests/src/test/java/its/SonarQubeDeveloperEditionTests.java]: addressed in [4989ca3](https://github.com/SonarSource/sonarlint-core/pull/799/commits/4989ca37baf60cf950245a99149ed3079780e0fa)

File: its/tests/src/test/java/its/SonarQubeDeveloperEditionTests.java [hunk 9]
```diff
@@ -1504,6 +1444,15 @@ private static void analyzeMavenProject(String projectDirName, String projectKey
     analyzeMavenProject(ORCHESTRATOR, projectDirName, Map.of("sonar.projectKey", projectKey));
   }
 
+  private static void bindProject(String projectName, String projectKey, boolean bindingSuggestionDisabled) {
+    backend.getConfigurationService().didAddConfigurationScopes(new DidAddConfigurationScopesParams(
+      List.of(new ConfigurationScopeDto(CONFIG_SCOPE_ID, null, true, projectName, new BindingConfigurationDto(CONNECTION_ID, projectKey,
+        bindingSuggestionDisabled)))));
+    await().atMost(20, SECONDS).untilAsserted(() -> assertThat(didSynchronizeConfigurationScopes).contains(CONFIG_SCOPE_ID));
+    await().untilAsserted(() -> assertThat(clientLogs.stream().anyMatch(s -> s.getMessage().equals("Stored project analyzer configuration"))).isTrue());
```
Comment (henryju) [inline its/tests/src/test/java/its/SonarQubeDeveloperEditionTests.java:1452]: Please add a big TODO or FIXME mentioning that this is a temporary workaround until SLCORE-396 is fixed

**Review by henryju — APPROVED** [review]

**Pull request merged**